### PR TITLE
RocksDB Trace Analyzer

### PR DESCRIPTION
**Pull request by zhichao-cao**
A framework of trace analyzing for RocksDB

After collecting the trace by using the tool of [PR #3837](https://github.com/facebook/rocksdb/pull/3837). User can use the Trace Analyzer to interpret, analyze, and characterize the collected workload. 
**Input:**
1. trace file
2. Whole keys space file

**Statistics:**
1. Access count of each operation (Get, Put, Delete, SingleDelete, DeleteRange, Merge) in each column family. 
2. Key hotness (access count) of each one
3. Key space separation based on given prefix
4. Key size distribution
5. Value size distribution if appliable
6. Top K accessed keys
7. QPS statistics including the average QPS and peak QPS
8. Top K accessed prefix
9. The query correlation analyzing, output the number of X after Y and the corresponding average time 
    intervals

**Output:**
1. key access heat map (either in the accessed key space or whole key space)
2. trace sequence file (interpret the raw trace file to line base text file for future use)
3. Time serial (The key space ID and its access time)
4. Key access count distritbution
5. Key size distribution
6. Value size distribution (in each intervals)
7. whole key space separation by the prefix
8. Accessed key space separation by the prefix
9. QPS of each operation and each column family
10. Top K QPS and their accessed prefix range

**Test:**
1. Added the unit test of analyzing Get, Put, Delete, SingleDelete, DeleteRange, Merge
2. Generated the trace and analyze the trace

**Implemented but not tested (due to the limitation of trace_replay):**
1. Analyzing Iterator, supporting Seek() and SeekForPrev() analyzing
2. Analyzing the number of Key found by Get

**Future Work:**
1.  Support execution time analyzing of each requests
2.  Support cache hit situation and block read situation of Get

### CODE REVIEW

File: tools/trace_analyzer_tool.cc [hunk 1]
```diff
@@ -0,0 +1,1621 @@
...
+}
+
+// Insert the corresponding key statistics to the correct type
+// and correct CF, output the time-serial file if needed
+Status TraceAnalyzer::KeyStatsInsertion(const uint32_t& type,
+                                        const uint32_t& cf_id,
+                                        const std::string& key,
+                                        const size_t value_size,
+                                        const uint64_t ts) {
+  auto found_stats = ta_[type].stats.find(cf_id);
```
Comment (siying) [inline tools/trace_analyzer_tool.cc]: Can you move this close to where `found_stats` is used?

File: tools/trace_analyzer_tool.cc [hunk 2]
```diff
@@ -0,0 +1,1621 @@
...
+    new_stats.corre_output.resize(analyzer_opts_.corre_list.size());
+    if (analyzer_opts_.output_prefix_cut) {
+      std::map<std::string, uint32_t> tmp_io_map;
+      tmp_io_map[prefix] = 1;
+      new_stats.a_io_prefix_stats[time_in_sec] = tmp_io_map;
+    }
+    ta_[type].stats[cf_id] = new_stats;
+  } else {
+    found_stats->second.a_count++;
+    found_stats->second.a_key_size_sqsum += key.size()*key.size();
```
Comment (siying) [inline tools/trace_analyzer_tool.cc]: Run "make format".

File: tools/trace_analyzer_tool.cc [hunk 3]
```diff
@@ -0,0 +1,1621 @@
...
+    found_stats->second.a_key_size_sum += key.size();
+    found_stats->second.a_value_size_sqsum += value_size*value_size;
```
Comment (siying) [inline tools/trace_analyzer_tool.cc]: This may overflow if user has length of GBs. You probably should check overflow here. It's OK to just exit saying it's not supported.

File: tools/trace_analyzer_tool_imp.h [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+  ~AnalyzerOptions();
+
+  void SparseCorreInput(const std::string& in_str);
+};
+
+
+struct TraceStats {
+  uint32_t cf_id;
+  std::string cf_name;
+  uint64_t a_count;
```
Comment (siying) [inline tools/trace_analyzer_tool_imp.h]: What does "a_" mean? Analysis?

File: tools/trace_analyzer_tool.cc [hunk 5]
```diff
@@ -0,0 +1,1621 @@
...
+
+// create the output path of the files to be opened
+FILE* TraceAnalyzer::CreateOutputFile(const std::string& type,
+                                      const std::string& cf_name,
+                                      const std::string& ending) {
+  std::string path;
+  FILE* f_new;
+  path = output_path_ + "/" + analyzer_opts_.output_prefix + "-" + type + "-" +
+         cf_name + "-" + ending;
+  f_new = fopen(path.c_str(), "w");
```
[the remark this answers is not in the document]
Comment (zhichao-cao) [reply tools/trace_analyzer_tool.cc]: Sure, I will replay them.

File: tools/trace_analyzer_tool.cc [hunk 6]
```diff
@@ -0,0 +1,1621 @@
...
+#include "util/string_util.h"
+#include "util/trace_replay.h"
+
+namespace rocksdb {
+
+std::map<std::string, uint32_t> cfname_to_cfid = {{"default", 0},
+      {"__system__", 1}, {"cf_fbobj", 2}, {"cf_assoc_count", 3},
+      {"cf_fbobj_deleter", 4}, {"cf_assoc", 5}, {"rev:cf_assoc_id1_type", 6},
+      {"cf_assoc_deleter", 7}, {"rev:cf_assoc_deleter_id1_type", 8},
+      {"cf_fbobj_type_id", 9}};
```
[the remark this answers is not in the document]
Comment (zhichao-cao) [reply tools/trace_analyzer_tool.cc]: They has been removed from the latest version, they are just temporally used for the MySQL workload

File: tools/trace_analyzer_tool_imp.h (hunk 4 above)
Comment (zhichao-cao) [reply tools/trace_analyzer_tool_imp.h]: "a" prefix means "accessed", "w" prefix means "whole". Using the full name makes some variable too long. Should I change them?

File: tools/trace_analyzer.cc [hunk 7]
```diff
@@ -0,0 +1,21 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (zhichao-cao) [reply tools/trace_analyzer.cc]: Sure.

File: tools/trace_analyzer_tool_imp.h [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+  taMerge = 5,
+  taIter = 6
+};
+
+struct TraceUnit {
+  uint32_t type;
+  std::string key;
+  size_t value_size;
+  uint64_t ts;
+  uint32_t cf_id;
```
[the remark this answers is not in the document]
Comment (zhichao-cao) [reply tools/trace_analyzer_tool_imp.h]: I have added to "CMakeLists.txt" and "src.mk" to build it. However, since the trace_replay is not in the master yet, the trace_analyzer cannot be build. So I removed the change to that two files before I create the pull requests. Later on, I will add the change back.

File: tools/trace_analyzer_tool.cc [hunk 9]
```diff
@@ -0,0 +1,1621 @@
...
+
+  for (int i = 0; i < static_cast<int>(analyzer_opts_.corre_list.size()); i++) {
+    stats.corre_output[i].first += unit.v_corre[i].count;
+    stats.corre_output[i].second += unit.v_corre[i].total_ts;
+  }
+  return Status::OK();
+}
+
+
+// Process the statistics of IO
```
[the remark this answers is not in the document]
Comment (zhichao-cao) [reply tools/trace_analyzer_tool.cc]: Sure, in the latest version, it has been corrected to "Qps"

File: tools/trace_analyzer_tool.cc (hunk 3 above)
Comment (zhichao-cao) [reply tools/trace_analyzer_tool.cc]: Sure

File: include/rocksdb/trace_analyzer_tool.h [hunk 10]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) 2013-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+#pragma once
+
+namespace rocksdb {
+int trace_analyzer_tool(int argc, char** argv);
```
Comment (siying) [inline include/rocksdb/trace_analyzer_tool.h]: Why do we need such a definition under include/rocksdb/trace_analyzer_tool.h?

File: util/trace_replay.h [hunk 11]
```diff
@@ -32,6 +32,7 @@ enum TraceType : char {
   kTraceEnd = 2,
   kTraceWrite = 3,
   kTraceGet = 4,
+  kTraceIter = 5,
```
Comment (siying) [inline util/trace_replay.h]: Does it belong to another pR?

File: utilities/trace_analyzer_test.cc [hunk 12]
```diff
@@ -0,0 +1,708 @@
...
+// Copyright (c) 2012 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#ifndef ROCKSDB_LITE
+#ifndef GFLAGS
+#include <cstdio>
+int main() {
+    fprintf(stderr, "Please install gflags to run trace_analyzer test\n");
+      return 1;
```
Comment (siying) [inline utilities/trace_analyzer_test.cc]: Will "make format" fix the indenting? If not, manually fix it.

File: utilities/trace_analyzer_test.cc [hunk 13]
```diff
@@ -0,0 +1,708 @@
...
+#include "util/trace_replay.h"
+#include "utilities/trace_analyzer_tool_imp.h"
+
+namespace rocksdb {
+namespace {
+static const int kMaxArgCount = 100;
+static const size_t kArgBufferSize = 100000;
+
+bool ReadOneLine(std::istringstream* iss, SequentialFile* seq_file,
+                 std::string* output, bool* has_data, Status* result) {
```
Comment (siying) [inline utilities/trace_analyzer_test.cc]: Is it the same as the one in options/options_parser.cc? If it is, reuse that one instead of duplicate the code.

File: utilities/trace_analyzer_test.cc [hunk 14]
```diff
@@ -0,0 +1,708 @@
...
+    std::unique_ptr<WritableFile> whole_f;
+    std::string whole_path = test_path_ + "/0.txt";
+    ASSERT_OK(env_->NewWritableFile(whole_path, &whole_f, env_options_));
+    std::ostringstream whole;
+    whole << "0x61\n"
+          << "0x62\n"
+          << "0x63\n"
+          << "0x64\n"
+          << "0x65\n"
+          << "0x66\n";
```
Comment (siying) [inline utilities/trace_analyzer_test.cc]: What's the purpose of using this std::ostringstream, rather than
```
std::string whole_str = "\x61\n\x62\n\x63\n\x64\n\x65\n\x66\n";
```
?

File: utilities/trace_analyzer_tool.cc [hunk 15]
```diff
@@ -0,0 +1,1751 @@
...
+    {0, "get"},          {1, "put"},   {2, "delete"},  {3, "single_delete"},
+    {4, "range_delete"}, {5, "merge"}, {6, "iterator"}};
+
+// Transfer the Microsecond time to date time
+std::string TraceAnalyzer::MicrosdToDate(uint64_t time_in) {
+  time_t tx = static_cast<time_t>(time_in / 1000000);
+  int rest = static_cast<int>(time_in % 1000000);
+  std::string date_time(ctime(&tx));
+  date_time.pop_back();
+  date_time += " +: " + std::to_string(rest);
```
Comment (siying) [inline utilities/trace_analyzer_tool.cc]: Rather than std::to_string(), use the platform independent `ToString()` in "util/string_util.h".

File: utilities/trace_analyzer_tool.cc [hunk 16]
```diff
@@ -0,0 +1,1751 @@
...
+std::map<int, std::string> taIndexToOpt = {
+    {0, "get"},          {1, "put"},   {2, "delete"},  {3, "single_delete"},
+    {4, "range_delete"}, {5, "merge"}, {6, "iterator"}};
+
+// Transfer the Microsecond time to date time
+std::string TraceAnalyzer::MicrosdToDate(uint64_t time_in) {
+  time_t tx = static_cast<time_t>(time_in / 1000000);
+  int rest = static_cast<int>(time_in % 1000000);
+  std::string date_time(ctime(&tx));
+  date_time.pop_back();
```
Comment (siying) [inline utilities/trace_analyzer_tool.cc]: I don't understand this.

File: utilities/trace_analyzer_tool.cc [hunk 17]
```diff
@@ -0,0 +1,1751 @@
...
+std::map<std::string, int> taOptToIndex = {
+    {"get", 0},          {"put", 1},   {"delete", 2},  {"single_delete", 3},
+    {"range_delete", 4}, {"merge", 5}, {"iterator", 6}};
+
+std::map<int, std::string> taIndexToOpt = {
+    {0, "get"},          {1, "put"},   {2, "delete"},  {3, "single_delete"},
+    {4, "range_delete"}, {5, "merge"}, {6, "iterator"}};
+
+// Transfer the Microsecond time to date time
+std::string TraceAnalyzer::MicrosdToDate(uint64_t time_in) {
```
Comment (siying) [inline utilities/trace_analyzer_tool.cc]: We have existing code printing out date+time: https://github.com/facebook/rocksdb/blob/master/env/posix_logger.h#L106-L120

please consider to reuse the existing code.

File: utilities/trace_analyzer_tool.cc [hunk 18]
```diff
@@ -0,0 +1,1751 @@
...
+  std::string date_time(ctime(&tx));
+  date_time.pop_back();
+  date_time += " +: " + std::to_string(rest);
+  return date_time;
+}
+
+namespace {
+
+bool ReadOneLine(std::istringstream* iss, SequentialFile* seq_file,
+                 std::string* output, bool* has_data, Status* result) {
```
Comment (siying) [inline utilities/trace_analyzer_tool.cc]: Same question as the other ReadOneLine(). Is it the same as the one existing in the code base? If it is, don't copy/paste the code. Call the existing function.

File: utilities/trace_analyzer_tool.cc [hunk 19]
```diff
@@ -0,0 +1,1751 @@
...
+// other statistic result such as key size distribution, value size
+// distribution, these data structures are re-processed here.
+Status TraceAnalyzer::MakeStatistics() {
+  int ret;
+  Status s;
+  for (int type = 0; type < kTaTypeNum; type++) {
+    if (!ta_[type].enabled) {
+      continue;
+    }
+    for (auto& i : ta_[type].stats) {
```
Comment (siying) [inline utilities/trace_analyzer_tool.cc]: This `i` is confusing. Maybe call it `stat` or `record`?

File: utilities/trace_analyzer_tool.cc [hunk 20]
```diff
@@ -0,0 +1,1751 @@
...
+      if (FLAGS_output_key_stats || FLAGS_output_prefix_cut > 0) {
+        s = MakeStatisticKeyStatsOrPrefix(i.second);
+        if (!s.ok()) {
+          return s;
+        }
+      }
+
+      // output the access count distribution
+      if (FLAGS_output_access_count_stats && i.second.a_count_dist_f) {
+        for (auto& it : i.second.a_count_stats) {
```
Comment (siying) [inline utilities/trace_analyzer_tool.cc]: Similar comment. The naming `it` makes it hard to read.

File: utilities/trace_analyzer.cc [hunk 21]
```diff
@@ -0,0 +1,25 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+#ifndef ROCKSDB_LITE
+#ifndef GFLAGS
+#include <cstdio>
+int main() {
```
Comment (sagar0) [inline utilities/trace_analyzer.cc]: I think trace_analyzer should go under tools rather than utilities, as this would be similar to, say, ldb/db_bench/sst_dump and other tools.

File: utilities/trace_analyzer_tool_imp.h [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+};
+
+// write bach handler to be used for WriteBache iterator
+// when processing the write trace
+class TraceWriteHandler : public WriteBatch::Handler {
+ private:
+  TraceAnalyzer* ta_ptr;
+  std::string tmp_use;
+
+ public:
```
Comment (sagar0) [inline utilities/trace_analyzer_tool_imp.h]: nit: public declarations should come before private and protected.

File: utilities/trace_analyzer_tool.cc [hunk 23]
```diff
@@ -0,0 +1,1751 @@
...
+    }
+
+    total_requests_++;
+    end_time_ = trace.ts;
+    if (trace.type == kTraceWrite) {
+      total_writes_++;
+      c_time_ = trace.ts;
+      WriteBatch batch(trace.payload);
+      if (batch.HasBeginPrepare() && !batch.HasCommit()) {
+        continue;
```
Comment (sagar0) [inline utilities/trace_analyzer_tool.cc]: Can you add a comment here on why this needs to be done, so that it is clear to everyone?
I did not know until you told me that we would double count writes in transactional RocksDB if we don't do this.

File: utilities/trace_analyzer_tool.cc [hunk 24]
```diff
@@ -0,0 +1,1751 @@
...
+  if (op1 == 0 || op2 == 0) {
+    return 0;
+  }
+  if (port::kMaxUint64 / op1 < op2) {
+    return op1;
+  }
+  return (op1 * op2);
+}
+
+void DecodeCFAndKey(std::string& buffer, uint32_t* cf_id, Slice* key) {
```
Comment (sagar0) [inline utilities/trace_analyzer_tool.cc]: I guess I can later pull these common functions to a separate utils file to remove some code redundancy.

File: utilities/trace_analyzer_tool_imp.h [hunk 25]
```diff
@@ -0,0 +1,267 @@
...
+  TraceWriteHandler() { ta_ptr = nullptr; }
+  TraceWriteHandler(TraceAnalyzer* _ta_ptr) { ta_ptr = _ta_ptr; }
+  ~TraceWriteHandler() {}
+
+  virtual Status PutCF(uint32_t column_family_id, const Slice& key,
```
Comment (sagar0) [inline utilities/trace_analyzer_tool_imp.h]: Suggestion: Might be good to get rid of CF suffix in all these functions, as it is implicitly assumed that we will support only operations with CF.

File: utilities/trace_analyzer_tool.cc (hunk 23 above)
Comment (zhichao-cao) [reply utilities/trace_analyzer_tool.cc]: Yes, it is important to point out it. Otherwise most people have no clue why the writes are doubled

File: utilities/trace_analyzer_tool.cc (hunk 19 above)
Comment (zhichao-cao) [reply utilities/trace_analyzer_tool.cc]: I will replace all these to make them clear.

File: utilities/trace_analyzer_tool_imp.h [hunk 26]
```diff
@@ -0,0 +1,267 @@
...
+
+struct TraceUnit {
+  uint64_t ts;
+  uint32_t type;
+  uint32_t cf_id;
+  size_t value_size;
+  std::string key;
+};
+
+struct TypeCorre {
```
[the remark this answers is not in the document]
Comment (zhichao-cao) [reply utilities/trace_analyzer_tool_imp.h]: Sure, use Correlation looks better

File: tools/trace_analyzer_tool.cc [hunk 27]
```diff
@@ -0,0 +1,1719 @@
...
+DEFINE_int32(output_prefix_cut, 0,
+             "The number of bytes as prefix to cut the keys."
+             "if it is enabled, it will generate the following:\n"
+             "for accessed keys:\n"
+             "format:[acessed_keyid access_count num_keys ave_access prefix]\n"
+             "for whole key space keys:\n"
+             "format:[start_keyid_in_whole_keyspace prefix]\n"
+             "if 'output_qps_stats' is enabled, it will output:\n"
+             "format:[time_in_sec IO_num], [prefix qps_of_this_second].");
+DEFINE_bool(output_trace_sequence, false,
```
Comment (sagar0) [inline tools/trace_analyzer_tool.cc]: 1. May be rename it to something like `convert_to_human_readable_trace`? Otherwise I think people might have a hard time finding this option. 
2. Could you also specify the filename where the converted output goes to?

File: tools/trace_analyzer_tool.cc [hunk 28]
```diff
@@ -0,0 +1,1719 @@
...
+  if (!s.ok()) {
+    return s;
+  }
+  if (footer->type != kTraceEnd) {
+    return Status::Corruption("Corrupted trace file. Incorrect footer.");
+  }
+  return s;
+}
+
+Status TraceAnalyzer::ReadTraceRecord(Trace* trace) {
```
Comment (sagar0) [inline tools/trace_analyzer_tool.cc]: For future: ReadTraceRecord, ReadTraceHeader and ReadTraceFooter should actually be directly using the code from Replayer/TraceReader, for better encapsulation. We should remove the duplication in the future.

**Comment by siying** [comment]
I don't have any concern. It's good to go!

**Comment by zhichao-cao** [comment]
@siying @sagar0 Thanks for the reviews and suggestions!

**Comment by sagar0** [comment]
Forgot about this: Could you add a line to HISTORY.md?

**Comment by zhichao-cao** [comment]
@sagar0 Sure, I will add a short description about the trace_analyzer tool.
